Sources/PortalSwift/Provider/PortalProvider.swift

```diff
@@ -214,82 +177,255 @@ public class PortalProvider {
+      case .eth_uninstallFilter, .net_listening:
+        let rpcResponse = try decoder.decode(PortalProviderRpcBoolResponse.self, from: data)
+        if let rpcError = rpcResponse.error {
+          throw PortalRpcError(rpcError)
+        }
+        return PortalProviderResult(id: forId, result: rpcResponse)
+      case .eth_getFilterChanges, .eth_getFilterLogs, .eth_getLogs:
+        let rpcResponse = try decoder.decode(LogsResponse.self, from: data)
+        if let rpcError = rpcResponse.error {
+          throw PortalRpcError(rpcError)
+        }
+        return PortalProviderResult(id: forId, result: rpcResponse)
+      default:
+        let rpcResponse = try decoder.decode(PortalProviderRpcResponse.self, from: data)
+        if let rpcError = rpcResponse.error {
+          throw PortalRpcError(rpcError)
+        }
+        return PortalProviderResult(id: forId, result: rpcResponse)
       }
-    } else if isSignerMethod {
-      self.handleSigningRequest(payload: payloadWithId, connect: connect) { (result: Result<SignerResult>, id: String) in
-        if id == payloadWithId.id, !self.processedSignatureIds.contains(id) {
-          self.processedSignatureIds.append(id)
+    }
 
-          guard result.error == nil else {
-            return completion(Result(error: result.error!))
-          }
+    throw URLError(.badURL)
+  }
+
+  private func handleSignRequest(
+    _ onChainId: String,
+    withPayload: PortalProviderRequestWithId,
+    forId _: String,
+    onBlockchain: PortalBlockchain,
+    connect: PortalConnect? = nil
+  ) async throws -> PortalProviderResult {
+    guard try await self.getApproval(onChainId, forPayload: withPayload, connect: connect) else {
+      throw ProviderSigningError.userDeclinedApproval
+    }
+
+    let rpcUrl = try getRpcUrl(onChainId)
+    let payload = PortalSignRequest(method: withPayload.method, params: withPayload.params)
 
-          // Trigger `portal_signatureReceived` event
-          _ = self.emit(
-            event: Events.PortalSignatureReceived.rawValue,
-            data: RequestCompletionResult(
-              method: payloadWithId.method,
-              params: payloadWithId.params,
-              result: result,
-              id: id
-            )
-          )
-
-          // Trigger completion handler
-          return completion(Result(data: RequestCompletionResult(method: payloadWithId.method, params: payloadWithId.params, result: result, id: id)))
+    let signature = try await self.signer.sign(
+      onChainId,
+      withPayload: payload,
+      andRpcUrl: rpcUrl,
+      usingBlockchain: onBlockchain
+    )
+
+    return PortalProviderResult(id: withPayload.id, result: signature)
+  }
+
+  private func removeOnce(registeredEventHandler: RegisteredEventHandler) -> Bool {
+    !registeredEventHandler.once
+  }
+
+  /*******************************************
+   * Deprecated functions
+   *******************************************/
```

Comment:
Updated to use new async request function while honoring the completion handler contract.